arangod/RocksDBEngine/RocksDBSyncThread.cpp

```diff
@@ -0,0 +1,140 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2017 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBSyncThread.h"
+#include "Basics/ConditionLocker.h"
+#include "Basics/RocksDBUtils.h"
+#include "Logger/Logger.h"
+#include "RocksDBEngine/RocksDBEngine.h"
+
+#include <rocksdb/status.h>
+#include <rocksdb/utilities/transaction_db.h>
+
+using namespace arangodb;
+
+RocksDBSyncThread::RocksDBSyncThread(RocksDBEngine* engine, std::chrono::milliseconds interval)
+    : Thread("RocksDBSync"), 
+      _engine(engine), 
+      _interval(interval),
+      _lastSyncTime(std::chrono::steady_clock::now()),
+      _lastSequenceNumber(0) {}
+
+RocksDBSyncThread::~RocksDBSyncThread() { shutdown(); }
+ 
+Result RocksDBSyncThread::syncWal() { 
+  // note the following line in RocksDB documentation (rocksdb/db.h):
+  // > Currently only works if allow_mmap_writes = false in Options.
+  TRI_ASSERT(!_engine->rocksDBOptions().allow_mmap_writes);
+      
+  auto db = _engine->db()->GetBaseDB();
+   
+  // set time of last syncing under the lock
+  auto const now = std::chrono::steady_clock::now();
+  {
+    CONDITION_LOCKER(guard, _condition);
+
+    if (now > _lastSyncTime) {
+      // update last sync time...
+      _lastSyncTime = now;
+    }
+        
+    auto lastSequenceNumber = db->GetLatestSequenceNumber();
+  
+    if (lastSequenceNumber > _lastSequenceNumber) {
+      // update last sequence number
+      _lastSequenceNumber = lastSequenceNumber;
+    }
+  }
+
+  // actual syncing is done without holding the lock
+  return sync(db);
+}
+  
+Result RocksDBSyncThread::sync(rocksdb::DB* db) {
+  LOG_TOPIC(TRACE, Logger::ROCKSDB) << "syncing RocksDB WAL";
```

Comment:
The code above contained a "#ifndef WIN32". Is sync now working under Windows?

Reply:
Good question. Chances are that never RocksDB versions actually support it.
We will test it once merged, and if it's still unsupported, we will fall back to the old behavior.